### ISSUE TITLE
Use determined.ai for active learning

### ISSUE THREAD

**Issue by MalteEbner** [determined-ai/determined#2888]
How can I use determined.ai for active learning? E.g. I want to conduct an experiment/episode consisting of multiple steps of training a machine learning model. The pseudocode is similar to

```python
model = Model(hyperparameters)
dataset = Dataset(hyperparameters)
validation_set = ValidationDataset()
active_learning_agent = ActiveLearningAgent(hyperparameters.agent_type)

# The model in one iteration depends on the model from the previous iterations, thus these are not independent.
for ratio in [0.1, 0.2, 0.3, 0.4, 1.0]:
    dataset_subset = active_learning_agent.query_subset(dataset, model, ratio)
    model.train(dataset_subset)
    model.evaluate(validation_set) # saves many metrics, e.g. the accuracy
```
After trying different active learning agents, I would like to generate a plot similar to the following. (the y-axis shows the classification accuracy)
<img width="397" alt="image" src="https://user-images.githubusercontent.com/20324507/131529433-29aca18a-2fd5-4cdc-96c6-3cfa8750fb4a.png">

**Comment by liamcli** [comment]
Hi Malte, 
You can repurpose our trials to do active learning runs but it's somewhat hacky.  One possible approach is to have each `train_batch` call be an active learning round where you prepare the labeled dataset, train for some number of epochs, and then compute whatever you need to identify points to label in the next active learning round.  I've provided a template trial definition and an experiment config below as an example.  Let me know if you have followup questions, I'm happy to help.  

```python
from typing import Any, Dict, Sequence, Union

import math
import os

import attrdict
import torch
from torchvision import transforms, datasets

import determined.pytorch as det_torch

TorchData = Union[Dict[str, torch.Tensor], Sequence[torch.Tensor], torch.Tensor]


class ActiveLearningTrial(det_torch.PyTorchTrial):
    def __init__(self, context: det_torch.PyTorchTrialContext) -> None:
        self.context = context
        self.hparams = attrdict.AttrDict(self.context.get_hparams())
        self.data_config = attrdict.AttrDict(self.context.get_data_config())

        # Model
        self.model = MyModel(self.hparams)
        self.model = self.context.wrap_model(self.model)

        self.optimizer = ... # build optimizer
        self.optimizer = self.context.wrap_optimizer(self.optimizer)

    def build_training_data_loader(self) -> Any:
        # This can be a counter dataset that just has indices corresponding to the rounds of
        # active learning that you want to do.
        train_loader = det_torch.DataLoader(
            list(range(self.hparams.active_learning_rounds)),
            batch_size=1,
            shuffle=False,
        )
        return train_loader

    def build_validation_data_loader(self) -> Any:
        # Build the validation data_loader as you would normally.
        val_dataset = ...
        val_loader = det_torch.DataLoader(
            val_dataset,
            batch_size=self.context.get_per_slot_batch_size(),
            num_workers=self.data_config.num_workers,
        )
        return val_loader

    def build_labeled_data_loader(self, ...) -> Any: 
        # Build the labeled dataset that you want to use to train the model for a given round.
        # You can include args to configure labeled dataset.  
        # You will probably need to also use instance attributes or something to track the
        # samples you want to label for the next round to create the next labeled dataset.
        labeled_dataset = ...

        return det_torch.DataLoader(
            labeled_dataset,
            batch_size=self.context.get_per_slot_batch_size(),
            num_workers=self.data_config.num_workers,
            shuffle=True,
            drop_last=True,
        )

    def train_batch(
        self, batch: TorchData, epoch_idx: int, batch_idx: int
    ) -> Dict[str, torch.Tensor]:
        # The batch provided will just be the active learning round index.
        # The logic here should include all the training you want to do 
        # for a given round of active learning.

        # We are going to create a separate dataloader for training the classifier in each round.
        self.labeled_dataloader = self.build_labeled_data_loader(...).get_data_loader(
            repeat=False,
            rank=self.rank,
        )
        for e in range(self.hparams.labeled_train_epochs):
            for i, batch in enumerate(self.labeled_dataloader):
                preds = self.model(batch)
                loss = ...
                self.context.backward(loss)
                self.context.step_optimizer(self.optimizer)
        # After training is finished for a round, you can use the trained model
        # to determine which datapoints should be labeled in the next round 
        # and add it to self.labeled_dataset.
        ...

        # For metrics, you may want to aggregate losses across batches during the training
        # loop or compute statistics on uncertainty etc. for unlabeled data that you pass 
        # through the model to determine what to label next.  
        ...

        return {metrics}

    def evaluate_batch(self, batch: TorchData) -> Dict[str, Any]:
        preds = self.model(batch)
        metrics = ...
        return {metrics}
```

along with an exp config:
```
description: active_learning_trial

hyperparameters:
  global_batch_size: 1
  active_learning_rounds: 10
  # How many epochs to loop over the labeled data when training during an active learning round.
  labeled_train_epochs: 10
  # Other hparams
  ...

data:
  num_workers: 8

min_validation_period:
  batches: 1
scheduling_units: 1 # Run 1 batch before returning training metrics.
searcher:
  name: single
  metric: accuracy
  smaller_is_better: false
  max_length:
    batches: 10 # this is how many rounds you want to run active learning
entrypoint: active_learning_example:ActiveLearningTrial
```

**Comment by MalteEbner** [comment]
Thank you for your quick answer! 
Did I get the following right?
- Basically, you are abusing the `train_batch()` method. It does not train on a single batch anymore, but on all batches for multiple epochs to perform one active learning step.
Thus the information during training of one single model is basically lost. Furthermore, there is less fine-grained control, as each "train_batch()" can easily run multiple hours.
- The saving of the model from the previous iteration has to be done manually.

**Comment by MalteEbner** [comment]
Thanks a lot for your help, I will try the 2nd option!

**Comment by shiyuann** [comment]
Hi @MalteEbner we are releasing a feature called custom searcher. With this, you can easily coordinate between trials within an experiment. You should be able to easily implement what you are trying to do here within a `SearchRunner` without hacking the product. Let me know if you want to know more about it! Thanks!
